### ISSUE TITLE
`jsx-no-useless-key` New rule to check for unnecessary use of the `key` prop - when not used in an iterator.

### ISSUE THREAD

**Issue by olyop** [jsx-eslint/eslint-plugin-react#3378]
Feature reqauest for to add a check for the `key` prop when it has been used unnecessarily.

I'm thinking a new rule would make the most sense but we could also add it to the rule `react/jsx-key` but that rule is releated to iterators which this feature is not only applicable to.

I may be missing a use case for this though.

❌ Incorrect
```JSX
export default () => (
  <div>

    {/* Unnecessary:
        Elements are static */}
    <div key="asdf"/>

    <div className="SomeList">
      {/* Unnecessary:
          Elements are static like this too, React can determine the order */}
      <p key="1">asdf</p>
      <p key="2">asdf</p>
      <p key="3">asdf</p>
    </div>

  </div>
)

/* Unnecessary:
   The `key` prop only needs to be defined when creating an instance
   of the component as shown in the correct section */
const User = ({ key, user }) => (
  <div key={key}>
      // ...
  </div>
)
```

✅ Correct
```JSX
export default ({ users }) => (
  <div>
    {users.map(user => (
      <User
        key={user.userID}
        user={user}
      />
    ))}
  </div>
)
```

I'm happy to try implement this myself and make the pull request but I'm not sure how to setup a local environment so maybe someone can help me with this.

**Comment by ljharb** [comment]
Note that the `key` prop does have a use outside of iterators - see https://kentcdodds.com/blog/understanding-reacts-key-prop

**Comment by olyop** [comment]
Thanks for confirming @ljharb I thought was the case.
How about a new rule with name `react/jsx-no-useless-key` that checks for use of the `key` prop when not used with an iterator.

**Comment by ljharb** [comment]
You may want to read the article :-) there is a use when it's not used with an iterator. It forces React to remount an element.

**Comment by olyop** [comment]
You're right, I must of misread you're comment before but it looks like that is a use case. Seems like an edge case to me or at least unconvential.

**Comment by ljharb** [comment]
It's not common, that's true - but there's no way to statically know if it's being used for that purpose, or for no purpose at all, which means it'd be inappropriate for a lint rule to warn on it.
